docs/platform_developers/database.rst

```diff
@@ -46,5 +49,65 @@ updates:
 3. When schema updates are available, each ``update`` function between the current version and the latest version is executed
    in the right order.
 
-When a schema update fails, the database schema is rolled-back to the latest schema version for which the ``update`` function
-did succeed. In that case the Inmanta server will fail to start.
+When a schema update fails, the database schema is rolled-back to the state before the start of the Inmanta server. In
+that case the Inmanta server will fail to start.
+
+
+Testing database migrations
+###########################
+
+Each database migration script should be tested using an automated test case. The tests that verify the migration from
+schema version ``<old_version>`` to ``<new_version>`` are stored in a file named
+``tests/db/test_v<old_version>_to_v<new_version>.py``.
+
+In general, a database schema migration test has the following flow:
+
+1. Load a database dump that uses the database schema version directly preceding the version being tested.
+2. Perform assertions that verify the database schema before the migration.
+3. Start the inmanta server to trigger the database migration scripts.
+4. Perform assertions to verify that the migration was done correctly.
+
+The example below shows a test for the above-mentioned database migration script.
+
+.. code-block:: python
+    :linenos:
+
+    # File: tests/db/test_v202101010_to_v202102220.py
+    @pytest.mark.db_restore_dump(os.path.join(os.path.dirname(__file__), "dumps/v202101010.sql"))
+    async def test_add_new_column_to_test_table(
+        migrate_db_from: abc.Callable[[], abc.Awaitable[None]],
+        get_columns_in_db_table: abc.Callable[[str], list[str]],
```

Comment:
This is technically not compatible with iso4 because it is Python 3.9+ (the `collections.abc` module has been updated quite a bit). I don't think we should change this though: it's a very minor thing, in a document that will probably only be used internally, for a product version that will become unsupported in 3 months. Do you agree this is acceptable?

Reply:
I agree.